crates/verify/src/bytecode/zksync.rs

```diff
@@ -0,0 +1,485 @@
+    VerifyBytecodeArgs,
+};
+use alloy_primitives::Address;
+use alloy_provider::Provider;
+use alloy_rpc_types::{BlockId, BlockNumberOrTag};
+use eyre::{OptionExt, Result};
+use foundry_block_explorers::{errors::EtherscanError, Response, ResponseData};
+use foundry_cli::utils::{self, LoadConfig};
+use foundry_common::{compile::ProjectCompiler, shell};
+use foundry_compilers::artifacts::{serde_helpers, EvmVersion};
+use foundry_zksync_compilers::compilers::zksolc::settings::{
+    BytecodeHash, Codegen, Optimizer, ZkSettings,
+};
+use reqwest::Url;
+use serde::{Deserialize, Deserializer, Serialize};
+use yansi::Paint;
+
+/// Run verify-bytecode for zksync. Since there's only runtime bytecode and that bytecode
+/// cannot be modified at deploy time by the constructor, the check is much simpler than
+/// in EVM as we do not need to simulate the deployment. We just compare the compiled bytecode
+/// against the on-chain one.
+pub async fn run(args: VerifyBytecodeArgs) -> Result<()> {
+    let config = args.load_config()?;
+    let provider = utils::get_provider(&config)?;
+    // If chain is not set, we try to get it from the RPC.
+    // If RPC is not set, the default chain is used.
+    let chain = match config.get_rpc_url() {
+        Some(_) => utils::get_chain(config.chain, &provider).await?,
+        None => config.chain.unwrap_or_default(),
+    };
+
+    // Get etherscan values.
+    let etherscan_config = config.get_etherscan_config_with_chain(Some(chain))?;
+    let etherscan_key = etherscan_config.as_ref().map(|c| c.key.as_str());
+    // TODO: we just create etherscan client to get the api url which has some
+    // resolution logic baked in. We cannot reuse the client for now as most primitive methods
+    // are private and we need zksync specific deserialization.
+    let etherscan = EtherscanVerificationProvider.client(
+        chain,
+        &args.verifier.verifier,
+        args.verifier.verifier_url.as_deref(),
+        etherscan_key,
+        &config,
+    )?;
+    let etherscan_api_url = etherscan.etherscan_api_url();
+
+    let onchain_runtime_code = match args.block {
+        Some(BlockId::Number(BlockNumberOrTag::Number(block))) => {
+            provider.get_code_at(args.address).block_id(BlockId::number(block)).await?
+        }
+        Some(_) => eyre::bail!("Invalid block number"),
+        None => provider.get_code_at(args.address).await?,
+    };
+
+    if onchain_runtime_code.is_empty() {
+        eyre::bail!("No bytecode found at address {}", args.address);
+    }
+
+    if !shell::is_json() {
+        sh_println!(
+            "Verifying bytecode for contract {} at address {}",
+            args.contract.name,
+            args.address
+        )?;
+    }
+
+    let mut json_results: Vec<JsonResult> = vec![];
+
+    let block_explorer_metadata =
+        block_explorer_contract_source_code(etherscan_api_url, etherscan_key, args.address).await?;
+
+    let onchain_bytecode_hash = block_explorer_metadata
```

Comment:
Is this actually a bytecode hash or the bytecode metadata hash or the hashing type (keccak)?